core/Rename.cpp

```diff
@@ -40,7 +40,28 @@ namespace olympia
             registerConsumerHandler(CREATE_SPARTA_HANDLER_WITH_DATA(Rename, creditsDispatchQueue_, uint32_t));
         in_reorder_flush_.
             registerConsumerHandler(CREATE_SPARTA_HANDLER_WITH_DATA(Rename, handleFlush_, FlushManager::FlushingCriteria));
+        in_rename_retire_ack_.registerConsumerHandler
+                (CREATE_SPARTA_HANDLER_WITH_DATA(Rename, getAckFromROB_, InstPtr));
         sparta::StartupEvent(node, CREATE_SPARTA_HANDLER(Rename, setupRename_));
+        
+        auto setup_map = [this] (core_types::RegFile reg_file, const uint32_t num_renames) {
+                        for(uint32_t i = 0; i < num_renames; ++i) {
```

Comment:
I _think_ this is a bug.  On line 97 below, you initialize the first 32 registers as already ready (0->31).  But here the free renames _start_ at 0 and will _include_ those renames already set as ready.  In essence they are already in use (and not free).  I think this is what you want:

```suggestion
                        sparta_assert(num_regs < num_renames);
                        for(uint32_t i = num_regs; i < num_renames; ++i) {
```

Reply:
If we count the marking of registers as ready at line 97 and never add them to the freelist at the beginning, then we will never access those "ready" registers, because we only push back into the freelist register destinations that are retired. 
Are we treating those registers marked as ready at line 97 as the initial ARFs, and then as we progress through renaming, the ARFs get mapped to a new PRF? We would then need a way to track what is the current "valid" ARF value and then when a current instruction with a destination to a register is finished, we can then retire the "valid" ARF register for that one.

Reply:
I don't think there's need to track ARFs separately...

I might have missed this in your implementation -- do you already have a mapping for those first 32 registers?  If not, then this might the difference in our thinking.  I have a suggestion for fix below.

Consider this example as the very first instruction in the trace:
```
     add x1, x2, x3
```
Now, in a real machine, if this were the first instruction, you're technically already borked -- what are the values for `x2` and `x3`?  They've never been written to!  In some implementations, these registers are mapped to either a null register or x0.  But in the perf model, you've already accounted for these registers by setting `x0 -> x31` as ready on line 97 (implying they've been written to).   Your register map should have that mapping reflected as well:
```
    for(uint32_t i = 0; i < num_regs; ++i) {
        map_table_[reg_file][i] = {i, false};
    }
```

After renaming (and my suggested fixes):
```
    add x1, x2, x3  -> add x1 -> x32, x2 -> x2, x3 -> x3
```
When the `add` instruction retires, `x1` will be placed onto the freelist.

But you make an interesting observation -- if I ask for 48 renames, does that include the original 32?  Technically, if the map were setup so that x0 -> x0 (not real, BTW), x1 -> x1, etc, then you actually have 48 - 32 or 16 actual "free" renames for use.

Reply:
Implemented in [1bf68f4](https://github.com/riscv-software-src/riscv-perf-model/pull/40/commits/1bf68f4f79753f81792e51ef39c2b06e261e82de)